psutil/arch/linux/users.c

```diff
@@ -5,13 +5,151 @@
  */
 
 #include <Python.h>
+#include <dlfcn.h>
 #include <utmp.h>
 
 #include "../../_psutil_common.h"
 
+/* Systemd function signatures that will be loaded */
+int (*sd_booted)(void);
+int (*sd_get_sessions)(char ***);
+int (*sd_session_get_leader)(const char *, pid_t *);
+int (*sd_session_get_remote_host)(const char *,char **);
+int (*sd_session_get_start_time)(const char *, uint64_t *);
+int (*sd_session_get_tty)(const char *, char **);
+int (*sd_session_get_username)(const char *, char **);
+
+#define dlsym_check(__h, __fn) do {           \
+    __fn = dlsym(__h, #__fn);                 \
+    if (dlerror() != NULL || __fn == NULL) {  \
+        dlclose(__h);                         \
+        return NULL;                          \
+    }                                         \
+} while (0)
+
+static void *
+load_systemd() {
+    void *handle = dlopen("libsystemd.so.0", RTLD_LAZY);
+    if (dlerror() != NULL || handle == NULL)
+        return NULL;
+
+    dlsym_check(handle, sd_booted);
+    dlsym_check(handle, sd_get_sessions);
+    dlsym_check(handle, sd_session_get_leader);
+    dlsym_check(handle, sd_session_get_remote_host);
+    dlsym_check(handle, sd_session_get_start_time);
+    dlsym_check(handle, sd_session_get_tty);
+    dlsym_check(handle, sd_session_get_username);
+
+    if (! sd_booted()) {
+        dlclose(handle);
+        return NULL;
+    }
+    return handle;
+}
+
+PyObject *
+psutil_users_systemd(PyObject *self, PyObject *args) {
+    char **sessions_list = NULL;
+    PyObject *py_retlist = PyList_New(0);
+    PyObject *py_tuple = NULL;
+    PyObject *py_username = NULL;
+    PyObject *py_tty = NULL;
+    PyObject *py_hostname = NULL;
+    PyObject *py_user_proc = NULL;
+    double tstamp = 0.0;
+    pid_t pid = 0;
+    int sessions;
+    const char *session_id = NULL;
+    char *username = NULL;
+    char *tty = NULL;
+    char *hostname = NULL;
+    uint64_t usec = 0;
+    void *handle = load_systemd();
+
+    if (! handle)
+        return NULL;
```

Comment:
This will cause `SystemError: <function users_systemd> returned NULL without setting an exception`. You need to return None instead. You can do `Py_RETURN_NONE;` instead of `return NULL;`.

Also, you need to pay attention to `py_retlist`, which gets initialized before`load_systemd()`. You need to make sure both are cleared in case of error. Not tested, but a solution might be:

```c
PyObject *py_retlist;
void *handle; 
....

handle = load_systemd();
if (! handle)
     Py_RETURN_NONE;

py_retlist = PyList_New(0);
if (py_retlist == NULL)
    goto error;
```

Reply:
Very good point. Done